### PR TITLE
feat: expression builder poc using v4 structure

### PR DESCRIPTION
**Pull request by samisayegh**
PoC using the v4 api. Thoughts?

https://coveord.atlassian.net/browse/KIT-1131

### CODE REVIEW

File: packages/headless/src/utils/expression-builder-sapi-v4.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    | 'lowerThan'
+    | 'lowerThanOrEqual'
+    | 'greaterThan'
+    | 'greaterThanOrEqual'
+    | 'phoneticMatch'
+    | 'fuzzyMatch'
+    | 'regexMatch'
+    | 'wildcardMatch'
+    | 'differentThan';
+  value: FieldValue; // shouldn't this be an array?
```
Comment (samisayegh) [inline packages/headless/src/utils/expression-builder-sapi-v4.ts:66]: @acote-coveo, it seems like `value` is not an array in swagger. Is this intended? or just a mistake in the doc?

File: packages/headless/src/utils/expression-builder-sapi-v4.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+interface FieldExists {
+  type: 'fieldExists';
+  fieldName: string;
+}
+
+// Not sure how to use this one.
+interface MatchAll {
+  type: 'matchAll';
+}
```
Comment (samisayegh) [inline packages/headless/src/utils/expression-builder-sapi-v4.ts:93]: @acote-coveo, how should this object be used? It only has a `type` property in the doc, hmm.

Comment (acote-coveo) [reply packages/headless/src/utils/expression-builder-sapi-v4.ts:93]: It also have a value of type string.

File: packages/headless/src/utils/expression-builder-sapi-v4.ts (hunk 1 above)
Comment (samisayegh) [reply packages/headless/src/utils/expression-builder-sapi-v4.ts:66]: Ok, I was wondering if the intent was to specify multiple `fieldExpression` objects, one for each value. Supporting arrays would  help make the overall object less verbose 👍

File: packages/headless/src/utils/expression-builder-sapi-v4.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
+type QueryExpresion =
```
Comment (ThibodeauJF) [inline packages/headless/src/utils/expression-builder-sapi-v4.ts:1]:
```suggestion
type QueryExpression =
```

File: packages/headless/src/utils/expression-builder-sapi-v4.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  name: string;
+  value: QueryExpresion;
+}
+
+interface QuerySyntax {
+  type: 'querySyntax';
+  value: string;
+}
+
+function createExpressionBuilderV4(expression?: QueryExpresion) {
```
Comment (ThibodeauJF) [inline packages/headless/src/utils/expression-builder-sapi-v4.ts:116]: why is expression optional? Could we either make it obligatory or remove it, forcing users to use addExpression

Comment (samisayegh) [reply packages/headless/src/utils/expression-builder-sapi-v4.ts:116]: We can remove it, providing the functionality as a method where the parameter is required.
The goal is to allow developers to build on top of an existing expression. e.g.

```
const exp1 = `a AND b AND c`
const exp2 = `(exp1) OR d OR e`
```
So when building `exp2`, `exp1` would be passed, and the `build` method would concatenate the two.

**Pull request merged**